### PR TITLE
ZK Circuits

### PR DESCRIPTION
**Pull request by TtheBC01**
### Release Notes
[JIRA Link]()

#### Summary:
This PR introduces a "circuits" package to the Snickerdoodle mono-repo. The circuits packages contains provable code written in 01js, a native typescript library for zkSNARKS, for proving various things about user state in a privacy preserving fashion. 

#### Intended results:
- Functions from /packages/circuits should be callable from core package
- Provable code should run in reasonable amount of time so as not to deter end users

#### Potential Failures:
- Code could be to slow for practical use case
- Code might not work on mobile, hopefully it does

#### Relevant Metrics/Indicators:
- Proving time is the primary metric
- verification time should be an order of magnitude less that proving time.

#### Testing Notes:
- We need to test extensively on mobile form factors.

### CODE REVIEW

File: packages/circuits/src/circom/CircomUtils.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  static stringToPoseidonHash(input: string): PoseidonHash {
+    const fields = CircomUtils.stringToFields(input);
+
+    const sum = fields.reduce((acc, field) => {
+      return CircomUtils.moduloAdd(acc, field);
+    }, 0n);
+
+    return PoseidonHash(poseidon1([sum]));
+
+    if (fields.length == 1) {
```
Comment (okanzdmr) [inline packages/circuits/src/circom/CircomUtils.ts:70]: unreachable code block

File: packages/core/src/implementations/business/InvitationService.ts [hunk 2]
```diff
@@ -135,37 +127,36 @@ export class InvitationService implements IInvitationService {
...
-                  .map(() => EInvitationStatus.Accepted);
-              });
+            // This is a problem but we'll just add a new accepted invitation
+            return this.dataWalletUtils
+              .deriveOptInInfo(
+                invitation.consentContractAddress,
+                context.dataWalletKey,
+              )
+              .andThen((optInInfo) => {
+                return this.invitationRepo.addAcceptedInvitations([optInInfo]);
```
Comment (okanzdmr) [inline packages/core/src/implementations/business/InvitationService.ts:153]: in AcceptedInvitation object (InvitationforStorage) we are not storing any information related to optInInfo so the above call seems not necessary

File: packages/objects/src/businessObjects/Permission.ts
Comment (okanzdmr) [inline packages/objects/src/businessObjects/Permission.ts:1]: I guess we do not need this

File: packages/objects/src/businessObjects/versioned/InvitationForStorage.ts [hunk 3]
```diff
@@ -1,44 +1,26 @@
...
 import {
   VersionedObject,
   VersionedObjectMigrator,
 } from "@objects/businessObjects/versioned/VersionedObject.js";
-import { EVMContractAddress, TokenId } from "@objects/primitives/index.js";
+import { EVMContractAddress } from "@objects/primitives/index.js";
 import { PropertiesOf } from "@objects/utilities/index.js";
 
 export class InvitationForStorage extends VersionedObject {
   public static CURRENT_VERSION = 1;
```
Comment (okanzdmr) [inline packages/objects/src/businessObjects/versioned/InvitationForStorage.ts:9]: missin version bump and migration logic

File: packages/objects/src/businessObjects/versioned/PermissionForStorage.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import { OptInInfo } from "@objects/businessObjects/OptInInfo.js";
+import {
+  VersionedObject,
+  VersionedObjectMigrator,
+} from "@objects/businessObjects/versioned/VersionedObject.js";
+import { EPermissionType, EWalletDataType } from "@objects/enum";
```
Comment (okanzdmr) [inline packages/objects/src/businessObjects/versioned/PermissionForStorage.ts:6]: bad import

File: packages/objects/src/businessObjects/versioned/PermissionForStorage.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import { DataPermissions, Permission } from "@objects/index";
```
Comment (okanzdmr) [inline packages/objects/src/businessObjects/versioned/PermissionForStorage.ts:7]: bad import

File: packages/core/src/implementations/business/InvitationService.ts (hunk 2 above)
Comment (moahammadalt) [reply packages/core/src/implementations/business/InvitationService.ts:154]: The method is calling this.persistence.updateRecord with the optInInfo, not sure if we really need it or not.

**Review by orhunerd — APPROVED** [review]

**Review by moahammadalt — APPROVED** [review]

**Review by lucas-sdl — APPROVED** [review]

**Pull request merged**